File: doubly_linked_list.js
// This doubly linked list implementation follows the following guide: https://hackernoon.com/the-little-guide-of-linked-list-in-javascript-9daf89b63b54

// Creating a separate scope for declaring variables 
{
    // Initializing linked list
    function LinkedList() {
      this.head = null;
      this.tail = null;
    }
    
    function Node(value, next, prev) {
      this.value = value;
      this.next = next;
      this.prev = prev;
    }

    // Adding required functional
    LinkedList.prototype.addToHead = function(value) {
      const newNode = new Node(value, this.head, null);
      if (this.head) this.head.prev = newNode;
      else this.tail = newNode; 
      this.head = newNode;
    };

    LinkedList.prototype.addToTail = function(value) {
      const newNode = new Node(value, null, this.tail);
      if (this.tail) this.tail.next = newNode;
      else this.head = newNode;
      this.tail = newNode;
    }

    LinkedList.prototype.removeHead = function() {
    if (this.head) {
        if (this.head.next) {
            this.head = this.head.next
            this.head.prev = null
        } else {
            this.head = null
            this.tail = null
        }
     }
    }

    LinkedList.prototype.removeTail = function () {
        if (this.tail) {
            if (this.tail.prev) {
                this.tail = this.tail.prev
                this.tail.next = null
            } else {
                this.head = null
                this.tail = null
            }
        }
    }

    // search from head
    LinkedList.prototype.search = function (searchValue) {
        let currentNode = this.head
        while (currentNode) {
            if (currentNode.value == searchValue) return currentNode;
            else currentNode = currentNode.next 
        }
    }

    LinkedList.prototype.searchBackwards = function (searchValue) {
        let currentNode = this.tail
        while (currentNode) {
            if (currentNode.value == searchValue) return currentNode
            else currentNode = currentNode.prev
        }
    }


    // Simple test of the implementation
    const list = new LinkedList();
    list.addToHead(300);
    list.addToHead(200);
    list.addToHead(100);

    list.addToTail(400);
    list.addToTail(500);
    list.addToTail(600);

//     list.removeHead()
//     list.removeTail()
//     console.log(list);
//     console.log(list.head.next.value);
//     searchResult = list.search(300)

    searchResult = list.searchBackwards (300)
    console.log("Search results: ")
    console.log(searchResult)
}

File: doubly_linked_list.py
# This doubly_linked_list.py is an implementation of doubly linked list based on the idea from CLRS, Chapter 10.2

class Linked_list:
    def __init__(self):
        self.head = None
        self.tail = None

    def node(self, key, prev=None, next=None):
        return {"key": key,
                "prev": prev,
                "next": next}

    def list_insert(self, key, start_from_head=True):
        if self.head:
            if start_from_head:
                new_node = self.node(key, None, self.head)
                self.head["prev"] = new_node
                self.head = new_node
            else:
                new_node = self.node(key, self.tail, None)
                self.tail["next"] = new_node
                self.tail = new_node

        else:
            new_node = self.node(key)
            self.head = new_node
            self.tail = self.head

    def list_search(self, k, start_from_head=True):
        if start_from_head:
            method = "next"
            node = self.head
        else:
            method = "prev"
            node = self.tail
        while node is not None and node["key"] != k:
            node = node[method]
        return node

    def list_delete(self, k):
        node = self.list_search(k)
        node_clone = node.copy()
        if node:
            if node["prev"] is not None:
                node["prev"]["next"] = node["next"]
            else:
                self.head = node["next"]
            if node["next"] is not None:
                node["next"]["prev"] = node["prev"]
        return node_clone

    def list_traverse(self, start_from_head=True):
        if start_from_head:
            print("> List traverse starting from head:")
            method = "next"
            node = self.head
        else:
            print("> List traverse starting from tail:")
            method = "prev"
            node = self.tail
        while node is not None:
            print(node)
            node = node[method]


# Running simple examples
print("\n::::: myList :::::")
myList = Linked_list()
myList.list_insert(1)
myList.list_insert(2)
myList.list_insert(3)
myList.list_insert(4)
myList.list_insert(5)
print("myList.head[key] = " + str(myList.head["key"]))
print("myList.list_search(3): ")
print(myList.list_search(3))
myList.list_search(3)["prev"] = None
print("myList.list_search(3): ")
print(myList.list_search(3))
print('myList.list_search(22)')
print(myList.list_search(22))

print("\n::::: mylist1 :::::")
myList1 = Linked_list()
myList1.list_insert(1)
myList1.list_insert(2)
myList1.list_insert(3)
myList1.list_insert(4)
myList1.list_insert(5)
myList1.list_traverse()
myList1.list_traverse(False)
print("myList1.list_search(3): ")
print(myList1.list_search(3))
myList1.list_delete(3)
myList1.list_traverse()
myList1.list_traverse(False)

print("myList1.list_search(3): ")
print(myList1.list_search(3))
print("myList1.list_search(2): ")
print(myList1.list_search(2))
print("myList1.list_search(2) starting from tail: ")
print(myList1.list_search(2, False))

print("\n::::: mylist2 :::::")
myList2 = Linked_list()
myList2.list_insert(1)
myList2.list_insert(2)
myList2.list_insert(3)
myList2.list_insert(4)
myList2.list_insert(5)
myList2.list_traverse()

print("\n::::: mylist3 :::::")
myList3 = Linked_list()
myList3.list_insert(1, False)
myList3.list_insert(2, False)
myList3.list_insert(3, False)
myList3.list_insert(4, False)
myList3.list_insert(5, False)
myList3.list_insert(7)
myList3.list_insert(7, False)
myList3.list_traverse()